lib/iris/coords.py

```diff
@@ -2402,8 +2402,9 @@ def _guess_bounds(self, bound_position=0.5, monthly=False, yearly=False):
                 points = np.empty(self.shape[0] + 2)
                 points[1:-1] = self.points
                 direction = 1 if self.points[-1] > self.points[0] else -1
-                points[0] = self.points[-1] - (self.units.modulus * direction)
-                points[-1] = self.points[0] + (self.units.modulus * direction)
+                modulus = np.float64(self.units.modulus)
```

Comment:
```suggestion
                modulus_type = np.promote_types(
                    self.points.dtype, type(self.units.modulus)
                )
                (modulus,) = np.array([self.units.modulus], dtype=modulus_type)
```

This is less elegant, but honouring the original dtype (e.g. if `self.points.dtype==np.float32`) should protect against unintended consequences.